aggregator/src/aggregator.rs

```diff
@@ -2751,375 +2650,481 @@ where
 /// See: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age
 const CORS_PREFLIGHT_CACHE_AGE: u32 = 24 * 60 * 60;
 
-/// Constructs a Warp filter that matches the request path for tasks, as well as whatever regex is
-/// provided, but without consuming the path or any values encoded into it, such that later filters
-/// can still extract path components.
-fn path_matching_filter(trailing_fragment_regex: &str) -> Result<BoxedFilter<()>, Error> {
-    let regex = Regex::new(&format!(
-        // Task ID is 32 bytes, hence 43 bytes of Base64URL
-        "tasks/[A-Za-z0-9-_]{{43}}/{trailing_fragment_regex}"
-    ))
-    .map_err(|e| Error::Internal(format!("error compiling regex: {e}")))?;
+/// Wrapper around a type that implements [`Encode`]. It acts as a Trillium handler, encoding the
+/// inner object and sending it as the response body, setting the Content-Type header to the
+/// provided media type, and setting the status to 200.
+struct EncodedBody<T> {
+    object: T,
+    media_type: &'static str,
+}
 
-    let filter = warp::path::peek()
-        .and_then(move |peek: warp::path::Peek| {
-            let regex = regex.clone();
-            async move {
-                if regex.is_match(peek.as_str()) {
-                    Ok(())
-                } else {
-                    Err(warp::reject::reject())
-                }
-            }
-        })
-        .untuple_one()
-        .boxed();
+impl<T> EncodedBody<T>
+where
+    T: Encode,
+{
+    fn new(object: T, media_type: &'static str) -> Self {
+        Self { object, media_type }
+    }
+}
 
-    Ok(filter)
+#[async_trait]
+impl<T> Handler for EncodedBody<T>
+where
+    T: Encode + Sync + Send + 'static,
+{
+    async fn run(&self, mut conn: Conn) -> Conn {
+        conn.headers_mut()
+            .insert(KnownHeaderName::ContentType, self.media_type);
+        conn.ok(self.object.get_encoded())
+    }
 }
 
-/// AggregationJobId is 16 bytes, hence 22 bytes of Base64URL
-const AGGREGATION_JOB_PATH_REGEX: &str = "aggregation_jobs/[A-Za-z0-9-_]{22}";
+/// A Trillium handler that checks for state set when sending an error response, and updates an
+/// OpenTelemetry counter accordingly.
+struct StatusCounter(Counter<u64>);
+
+impl StatusCounter {
+    fn new(meter: &Meter) -> Self {
+        Self(
+            meter
+                .u64_counter("janus_aggregator_responses_total")
+                .with_description(
+                    "Count of requests handled by the aggregator, by method, route, and response status.",
+                )
+                .init(),
+        )
+    }
+}
 
-/// CollectionId is 16 bytes, hence 22 bytes of Base64URL
-const COLLECTION_JOB_PATH_REGEX: &str = "collection_jobs/[A-Za-z0-9-_]{22}";
+#[async_trait]
+impl Handler for StatusCounter {
+    async fn run(&self, conn: Conn) -> Conn {
+        conn
```

Comment:
I'm intentionally not following that pattern here, because I want the code to run whether or not a preceding handler set the halt flag. (thus it will unconditionally bump the counter) In fact, I think the halt flag will be set by every handler inside the router, and only unset for 404s that didn't match any route. This handler isn't participating in construction of a response, so we don't need to coordinate with other handlers through the halt flag.

Reply:
I think the idea is that the handler's post-handling in `before_send` doesn't run if the handler halts _before_ the handler's initial handling in `run` occurs -- simulating a "wrapper" handler, I suppose. We definitely want to run `before_send` if the connection (expectedly) halts "inside" the `StatusCounter`.